File: 116.populating-next-right-pointers-in-each-node.py
#
# @lc app=leetcode id=116 lang=python3
#
# [116] Populating Next Right Pointers in Each Node
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None
        
        
        level_start = root
        curr_node = None

        while level_start.left:
            curr_node = level_start

            while curr_node:
                curr_node.left.next = curr_node.right
                if curr_node.next:
                    curr_node.right.next = curr_node.next.left
                curr_node = curr_node.next
            
            level_start = level_start.left
        
        return root
        
# @lc code=end

File: 116.populating-next-right-pointers-in-each-node.java
/*
 * @lc app=leetcode id=116 lang=java
 *
 * [116] Populating Next Right Pointers in Each Node
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        
        if (root == null) {
            return null;
        }

        Node levelStart = root;
        Node currentNode;

        while (levelStart.left != null) {
            currentNode = levelStart;

            while (currentNode != null) {
                currentNode.left.next = currentNode.right;

                if (currentNode.next != null) {
                    currentNode.right.next = currentNode.next.left;
                }

                currentNode = currentNode.next;
            }

            levelStart = levelStart.left;
        }

        return root;
    }
}
// @lc code=end

